lib/std/zig/Ast.zig

```diff
@@ -4142,3 +4142,625 @@ test {
     _ = Parse;
     _ = Render;
 }
+
+/// printNode prints a visual representation of the AST in a human-readable
+/// format, starting at `node`.
+///
+/// This funciton is expected to be used for debugging purposes.
+pub fn printNode(tree: *const Ast, w: *Writer, node: Node.Index) Writer.Error!void {
+    var p: Printer = .{
+        .tree = tree,
+        .w = w,
+    };
+    try p.printNode(node);
+    try p.w.writeByte('\n');
+}
+
+const Printer = struct {
+    tree: *const Ast,
+    w: *Writer,
+
+    depth: usize = 0,
+
+    fn printNamedOptionalToken(p: *Printer, name: []const u8, token: OptionalTokenIndex) Writer.Error!void {
+        if (token.unwrap()) |t| {
+            try p.printNamedToken(name, t);
+        } else {
+            try p.w.print("{s} = .none", .{name});
+        }
+    }
+
+    fn printNamedOptionalNode(p: *Printer, name: []const u8, node: Node.OptionalIndex) Writer.Error!void {
+        if (node.unwrap()) |t| {
+            try p.printNamedNode(name, t);
+        } else {
+            try p.w.print("{s} = .none", .{name});
+        }
+    }
+
+    fn printNamedToken(p: *Printer, name: []const u8, token: TokenIndex) Writer.Error!void {
+        const tok = p.tree.tokens.get(token);
+        const loc = p.tree.tokenLocation(0, token);
+        if (p.tree.tokenTag(token).lexeme() == null) {
+            try p.w.print("{s} = .{s} {s} ({}:{})", .{ name, @tagName(tok.tag), p.tree.tokenSlice(token), loc.line + 1, loc.column + 1 });
+        } else {
+            try p.w.print("{s} = .{s} ({}:{})", .{ name, @tagName(tok.tag), loc.line + 1, loc.column + 1 });
+        }
+    }
+
+    fn printNamedNode(p: *Printer, name: []const u8, node: Node.Index) Writer.Error!void {
+        try p.w.print("{s} = ", .{name});
+        try p.printNode(node);
+    }
+
+    fn indent(p: *const Printer) Writer.Error!void {
+        const tab: []const u8 = "\n" ++ (".  " ** 256);
```

Comment:
You might wonder why the tree is indented using `". "` (a dot followed by two spaces), which at first glance seems weird.

This comes from Go’s `go/ast` (see [print.go](https://github.com/golang/go/blob/ad3ccd92e4c2ddfc1499a5e038eb0aa0012c0dda/src/go/ast/print.go#L87)).
The reason is practical: using a visible dot makes it easier to spot the end braces and distinguish indentation levels in the printed tree